Python/ceval.c

```diff
@@ -7033,6 +7033,20 @@ PyEval_SetProfile(Py_tracefunc func, PyObject *arg)
     }
 }
 
+void
+PyEval_SetProfileAllThreads(Py_tracefunc func, PyObject *arg)
+{
+    PyThreadState *this_tstate = _PyThreadState_GET();
+    PyInterpreterState* interp = this_tstate->interp;
+    PyThreadState* ts = PyInterpreterState_ThreadHead(interp);
+    while (ts) {
+        if (_PyEval_SetProfile(ts, func, arg) < 0) {
+            _PyErr_WriteUnraisableMsg("in PyEval_SetProfileAllThreads", NULL);
+        }
+        ts = PyThreadState_Next(ts);
```

Comment:
When a Python thread completes, _PyThreadState_DeleteCurrent() is called. This function changes ``prev`` and ``next`` members of PyThreadState to remove the thread state from the double linked list. What happens if a thread completes while we iterate on that list? Is it possible that we miss a few threads?

In Python/pystate.c, ``HEAD_LOCK(runtime)`` and ``HEAD_UNLOCK(runtime)`` is used to update these ``prev`` and ``next`` members. Should we hold the lock here?

The problem is that _PyEval_SetProfile() can execute arbitrary Python code so we should release the lock to call it.

Maybe... only hold the lock just to call PyInterpreterState_ThreadHead() and PyThreadState_Next()?

How these functions should be used it not documented :-( See https://github.com/python/cpython/issues/40857 for details.

Reply:
> When a Python thread completes, _PyThreadState_DeleteCurrent() is called. This function changes prev and next members of PyThreadState to remove the thread state from the double linked list. What happens if a thread completes while we iterate on that list? Is it possible that we miss a few threads?

To call this function you need the GIL and to call `_PyThreadState_DeleteCurrent` you also need the GIL so there is no possibility of breaking anything because they cannot be called at the same time, no?

Reply:
Indeed, `_PyThreadState_DeleteCurrent ` releases the GIL when it finishes: https://github.com/python/cpython/blob/a613fedd6e18e4ab382cf81ec767e1135fc949a7/Python/pystate.c#L1113

Reply:
> Maybe... only hold the lock just to call PyInterpreterState_ThreadHead() and PyThreadState_Next()?

That cannot prevent the linked list to be changed in the middle of the `_PyEval_SetProfile` calls

Reply:
Ah, I think I understand some part of the problem now. I am pushing a commit adding locks around the whole loop.

Reply:
Check out [2efeb91](https://github.com/python/cpython/pull/93504/commits/2efeb9143efb7e6290a9c4e61b6001d8f35b430a)